php_phongo.c

```diff
@@ -245,6 +264,50 @@ static void phongo_cursor_init(zval *return_value, mongoc_cursor_t *cursor, mong
 	intern->cursor = cursor;
 	intern->server_id = mongoc_cursor_get_hint(cursor);
 	intern->client = client;
+
+	if (zreadPreference) {
+#if PHP_VERSION_ID >= 70000
+		ZVAL_ZVAL(&intern->read_preference, zreadPreference, 1, 0);
+#else
+		Z_ADDREF_P(zreadPreference);
+		intern->read_preference = zreadPreference;
+#endif
+	}
+} /* }}} */
+
+static void phongo_cursor_init_for_command(zval *return_value, mongoc_cursor_t *cursor, mongoc_client_t *client, const char *db, zval *zcommand, zval *zreadPreference TSRMLS_DC) /* {{{ */
+{
+	php_phongo_cursor_t *intern;
+
+	phongo_cursor_init(return_value, cursor, client, zreadPreference TSRMLS_CC);
+	intern = Z_CURSOR_OBJ_P(return_value);
+
+	intern->database = estrdup(db);
+
+#if PHP_VERSION_ID >= 70000
+	ZVAL_ZVAL(&intern->command, zcommand, 1, 0);
+#else
+	Z_ADDREF_P(zcommand);
+	intern->command = zcommand;
+#endif
+} /* }}} */
+
+static void phongo_cursor_init_for_query(zval *return_value, mongoc_cursor_t *cursor, mongoc_client_t *client, const char *namespace, zval *zquery, zval *zreadPreference TSRMLS_DC) /* {{{ */
+{
+	php_phongo_cursor_t *intern;
+
+	phongo_cursor_init(return_value, cursor, client, zreadPreference TSRMLS_CC);
+	intern = Z_CURSOR_OBJ_P(return_value);
+
+	/* namespace has already been validated by phongo_execute_query() */
+	phongo_split_namespace(namespace, &intern->database, &intern->collection);
```

Comment:
While we could pass in `database` and `collection` strings here, that would require copying the `efree()` lines for each of the 4-5 early return statements in `phongo_execute_query()`. It seemed easier to just call `phongo_split_namespace()` a second time here.